File: techextract/backend/src/routes/flashcards.ts
import { Router } from "express";
import {
  getAllQuestions,
  getQuestionsBySkills,
  getCategories,
  getQuestionById,
} from "../lib/questions";

const router = Router();

router.get("/", async (req, res): Promise<void> => {
  const { skills, category, limit = "20", offset = "0", shuffle = "true", difficulty } = req.query as Record<string, string>;

  let questions = skills
    ? getQuestionsBySkills(skills.split(",").map(s => s.trim().toLowerCase()))
    : getAllQuestions();

  if (category) {
    questions = questions.filter(q => q.category.toLowerCase() === category.toLowerCase());
  }

  if (difficulty) {
    questions = questions.filter(q => q.difficulty === difficulty);
  }

  if (shuffle !== "false") {
    questions = [...questions].sort(() => Math.random() - 0.5);
  }

  const total = questions.length;
  const off = parseInt(offset, 10) || 0;
  const lim = parseInt(limit, 10) || 20;
  const paginated = questions.slice(off, off + lim);

  res.json({ success: true, total, offset: off, limit: lim, questions: paginated });
});

router.get("/categories", async (_req, res): Promise<void> => {
  const categories = getCategories();
  res.json({ success: true, categories });
});

router.get("/stats", async (_req, res): Promise<void> => {
  const all = getAllQuestions();
  const cats = getCategories();
  const breakdown = cats.map(cat => ({
    category: cat,
    count: all.filter(q => q.category === cat).length,
  }));
  res.json({ success: true, total: all.length, breakdown });
});

router.get("/:id", async (req, res): Promise<void> => {
  const raw = Array.isArray(req.params["id"]) ? req.params["id"][0] : req.params["id"];
  const id = parseInt(raw ?? "", 10);
  if (isNaN(id)) {
    res.status(400).json({ error: "Invalid ID" });
    return;
  }
  const question = getQuestionById(id);
  if (!question) {
    res.status(404).json({ error: "Question not found" });
    return;
  }
  res.json({ success: true, question });
});

export default router;

File: backend/src/routes/flashcards.py
import random
from flask import Blueprint, request, jsonify
from ..models import Question
from ..extensions import db
<<<<<<< HEAD
from ..routes.auth import login_required
=======
>>>>>>> 32a8c637a5fefa3de0db62ac8676f127794edeb8
from sqlalchemy import func

flashcards_bp = Blueprint("flashcards", __name__)


@flashcards_bp.get("/")
<<<<<<< HEAD
@login_required
def get_flashcards(current_user):
=======
def get_flashcards():
>>>>>>> 32a8c637a5fefa3de0db62ac8676f127794edeb8
    skills_param = request.args.get("skills", "")
    category = request.args.get("category", "")
    difficulty = request.args.get("difficulty", "")
    limit = min(int(request.args.get("limit", 20)), 200)
    offset = int(request.args.get("offset", 0))
    shuffle = request.args.get("shuffle", "true").lower() != "false"

    query = Question.query

    if skills_param:
        skills_list = [s.strip().lower() for s in skills_param.split(",") if s.strip()]
        query = query.filter(Question.skill.in_(skills_list))

    if category:
        query = query.filter(func.lower(Question.category) == category.lower())

    if difficulty:
        query = query.filter(Question.difficulty == difficulty)

    total = query.count()

    if shuffle:
        query = query.order_by(func.random())
    else:
        query = query.order_by(Question.id)

    questions = query.offset(offset).limit(limit).all()

    return jsonify({
        "success": True,
        "total": total,
        "offset": offset,
        "limit": limit,
        "questions": [q.to_dict() for q in questions],
    })


@flashcards_bp.get("/categories")
<<<<<<< HEAD
@login_required
def get_categories(current_user):
=======
def get_categories():
>>>>>>> 32a8c637a5fefa3de0db62ac8676f127794edeb8
    rows = db.session.query(Question.category).distinct().order_by(Question.category).all()
    categories = [r[0] for r in rows]
    return jsonify({"success": True, "categories": categories})


@flashcards_bp.get("/stats")
<<<<<<< HEAD
@login_required
def get_stats(current_user):
=======
def get_stats():
>>>>>>> 32a8c637a5fefa3de0db62ac8676f127794edeb8
    total = Question.query.count()
    rows = (
        db.session.query(Question.category, func.count(Question.id))
        .group_by(Question.category)
        .order_by(Question.category)
        .all()
    )
    breakdown = [{"category": cat, "count": count} for cat, count in rows]
    return jsonify({"success": True, "total": total, "breakdown": breakdown})


@flashcards_bp.get("/<int:question_id>")
<<<<<<< HEAD
@login_required
def get_question(current_user, question_id: int):
=======
def get_question(question_id: int):
>>>>>>> 32a8c637a5fefa3de0db62ac8676f127794edeb8
    question = Question.query.get(question_id)
    if not question:
        return jsonify({"error": "Question not found"}), 404
    return jsonify({"success": True, "question": question.to_dict()})
